### ISSUE TITLE
All container jobs fail on 2.196.1 with `ENOENT`

### ISSUE THREAD

**Issue by samdoran** [microsoft/azure-pipelines-agent#3669]
### Have you tried troubleshooting?
Oh yes. A [very simple job](https://github.com/ansible/ansible/blob/2f15b7d3f39b6ea9b9c9f8c51aab120ad0e93975/.azure-pipelines/azure-pipelines.yml) with one step that runs one command [fails with the same error](https://dev.azure.com/ansible/ansible/_build/results?buildId=31780&view=logs&j=9d538be6-44f8-50c3-7930-4e7f88a0f6b2&t=d6e50871-e6ea-5bf9-8b6e-cc7564100bbd).

Running the job in the VM rather than in a container does not cause this error.

I did some inspecting of the VM and `/agent/externals/node10/bin` does exist.

`/agent` is bind mounted into the running container (`-v "/agent/externals":"/__a/externals":ro`) and then `/__a/externals/node/bin/node /__w/_temp/containerHandlerInvoker.js` is called and fails (see logs below).

### Agent Version and Platform
2.196.1
[Custom VM](https://github.com/ansible/azure-pipelines-test-vm/tree/main/image/templates) in a Scale Set based on Ubuntu 18.04
[Custom container image](https://github.com/ansible/azure-pipelines-test-container/blob/main/Dockerfile) based on Ubuntu 20.04

### Azure DevOps Type and Version
[dev.azure.com/ansible](https://dev.azure.com/ansible/ansible/_build?definitionId=20&_a=summary)

### What's not working?
Yesterday all jobs in our pipeline started failing with the following error:

<details>
  <summary>Failed job output</summary>

```terminal
Starting: Bash
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.195.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: spawn /__a/externals/node10/bin/node ENOENT
    at exports._errnoException (util.js:1020:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:197:32)
    at onErrorNT (internal/child_process.js:376:16)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
##[error]Exit code 1 returned from process: file name '/usr/bin/docker', arguments 'exec -i -u 1001  -w /home/AzDevOps_azpcontainer b4809c3e2c89450b0f0333751ffbea11549e883eee6689d9a7b831bd80166b3e /__a/externals/node/bin/node /__w/_temp/containerHandlerInvoker.js'.
Finishing: Bash
```
  
</details>

This seems very similar to https://github.com/microsoft/azure-pipelines-agent/issues/3400 and https://github.com/microsoft/azure-pipelines-agent/issues/3520, except we are not checking out multiple repos.

We made no changes to our VM image or container image, but this failure started happening yesterday. A while ago we removed the `unattended-upgrades` package from our VM image to prevent `docker` from being updated automatically, so the Docker version has not changed.

 I am somewhat confident this started happening as the VMs started using 2.196.1. I did see that [changes were made](https://github.com/microsoft/azure-pipelines-agent/commit/6a5fc2cab2cff5dc2014594117f3de1866a8c26b#diff-25a6634263c1b1f6fc4697a04e2b9904ea4b042a89af59dc93ec1f5d44848a26) to how processes running in a container are handled but I don't see a direct link between that and the failures we are seeing.

We have had _some_ successful runs with 2.196.1, but the overwhelming majority of the jobs run with 2.196.1 are failing with increasing consistency.

**Comment by nitzmahone** [comment]
Just some more color here- it looks like there's something that's changed between 2.195.2 and 2.196.1 that's introduced a race or some other new behavior that sometimes confuses the job numbers. When running container-based jobs, the container init step always mounts the `/1` work dir into the container, but the following checkout step sometimes moves the cloned repo into the `/2` job dir, which will always fail on container-based jobs since that jobdir isn't shared into the container. For host-only jobs, it probably works fine, since it can see all the jobdirs. Forcing our VMSS agent extension to install 2.195.2 has temporarily fixed the problem, but there's definitely something going on with newer versions of the agent.

**Comment by krsche** [comment]
@nitzmahone Thanks for the input, we're currently also struggling with this. For some reason one of our self-hosted agent-pools selects the pre-release version (2.196.1) while all others are on the latest release (2.195.2). 
How did you force 2.195.2 for the agents? I see settings in the extension, but I think this is configured by Azure DevOps automatically, right?

**Comment by sivel** [comment]
Here are some additional logs:

From "initialize containers" (reformatted for readability):

```
/usr/bin/docker create --name default_quayioansibleazurepipelinestestcontainer190_c48915 \
    --label b0e97f --network vsts_network_545ac4326a8c4790837fd9c0d884c176  \
    -v "/var/run/docker.sock":"/var/run/docker.sock" \
    -v "/agent/_work/2":"/__w/2" \
    -v "/agent/_work/_temp":"/__w/_temp" \
    -v "/agent/_work/_tasks":"/__w/_tasks" \
    -v "/agent/_work/_tool":"/__t" \
    -v "/agent/externals":"/__a/externals":ro \
    -v "/agent/_work/.taskkey":"/__w/.taskkey" \
    quay.io/ansible/azure-pipelines-test-container:1.9.0 \
    "/__a/externals/node/bin/node" -e "setInterval(function(){}, 24 * 60 * 60 * 1000);"
```

From "checkout":

```
##[debug]Processed: ##vso[plugininternal.updaterepositorypath alias=self;]/agent/_work/1/ansible
##[debug]Repository requires to be placed at '/agent/_work/1/ansible', current location is '/agent/_work/2/s'
Repository is current at '/agent/_work/2/s', move to '/agent/_work/1/ansible'.
```

You can see `-v "/agent/_work/2":"/__w/2"` from the container initialization, and then the checkout moves the checkout to `/agent/_work/1` which is not mounted in the container.

The ENOENT, seems to actually be coming from:

```
2021-12-16T22:17:06.7290131Z [pid   109] chdir("/__w/1/s")           = -1 ENOENT (No such file or directory)
2021-12-16T22:17:06.7290591Z [pid   109] write(19, "\376\377\377\377", 4) = 4
2021-12-16T22:17:06.7290948Z [pid   109] exit_group(127)             = ?
2021-12-16T22:17:06.7291370Z [pid   102] <... read resumed>"\376\377\377\377", 4) = 4
2021-12-16T22:17:06.7291763Z [pid   109] +++ exited with 127 +++
2021-12-16T22:17:06.7292721Z [pid   102] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=109, si_uid=1001, si_status=127, si_utime=0, si_stime=0} ---
```

**Comment by nitzmahone** [comment]
@krsche yes, the agent extension is configured automatically, but it's still just a normal resource... I haven't been able to get the Azure CLI to update that resource (it always wants to make a new one for some reason), but I was able to use [Resource Explorer](https://resources.azure.com/) directly against the REST API to change it to 2.195.2, flush the VMSS pool and verify that new instances come up on 2.195.2 and work properly.

It does overwrite my changes every hour or so though, so I'm working on a timer-triggered Azure Function that will just force it to 2.195.2 every few minutes to hopefully keep our pool reasonably clear of the new version of the agent.

**Comment by nitzmahone** [comment]
Also, we're fairly certain the breaking change was introduced in #3586 (thanks @sivel 
 for digging through the diffs!). It looks like there were some other bugs introduced by this change that have already been fixed (eg #3654), so maybe that'll fix this one as well, but hard to tell, since I don't want to package and publish my own agent from source to test it.

**Comment by krsche** [comment]
@nitzmahone we've figured out how to change the extension settings via the `azure cli vmss set`command. But the options for that command are not very clear, I have to double check which command line options we provided in the end, but it worked without recreating the extension. It just modified the downloadUrl in the settings to 2.195.2.

We also noticed that this is not permanent and gets overwritten after x minutes.

Why the extension is using 2.196.1 for some pools is still unclear to me since it's only in pre release.

**Comment by nitzmahone** [comment]
@krsche Where are you seeing that it's pre-release?

I'm guessing they probably just roll it out across different tenants/regions/scalesets or something. Looking at the activity log on the VMSS, it's getting forced back every 2h. My scheduled Azure function (I ended up just writing it using the new C# client library that's in preview) is running every 5min to force it back if it changes from the value we want- so far so good...

**Comment by krsche** [comment]
@nitzmahone just from the releases page here. It says latest for 2.195.2 and pre-release for 2.196.x see:
![Screenshot_20211218-025928__01__01.jpg](https://user-images.githubusercontent.com/5144985/146625401-4e261b12-ae87-4d6d-9ef8-10f2b7f1e04c.jpg)
I'm just assuming this is up to date and the main release channel?

**Comment by nitzmahone** [comment]
¯\_(ツ)_/¯ no idea, but glad to know we're being used as guinea pigs for busted agent code and paying handsomely for the privilege... :frowning_face:

**Comment by golubev-ivan** [comment]
Hi @samdoran, @nitzmahone, @sivel, @krsche, thanks for reporting this issue, we've stopped release process of agent 2.196.1 and rolled back this agent for affected organizations. Currently, all organizations should use agent 2.195.2. Let me close this item since the issue was mitigated. Feel free to ping me, if you still have any questions.

**Comment by jessehouwing** [comment]
Was this fixed in 2.196.2?

**Comment by golubev-ivan** [comment]
Hi @jessehouwing, yes, in 2.196.2 this problem was fixed, but due to deployment freeze this version of the agent will start the official release process after 10 Jan.
